tests/helpers/test_script.py

```diff
@@ -315,837 +243,511 @@ def heard_event_cb(event):
+)
+async def test_wait_template_timeout(
+    hass, mock_timeout, continue_on_timeout, n_events, script_mode
+):
+    """Test the wait template, halt on timeout."""
     event = "test_event"
-    wait_started_flag = asyncio.Event()
+    events = async_capture_events(hass, event)
+    sequence = [
+        {"wait_template": "{{ states.switch.test.state == 'off' }}", "timeout": 5},
+        {"event": event},
+    ]
+    if continue_on_timeout is not None:
+        sequence[0]["continue_on_timeout"] = continue_on_timeout
+    sequence = cv.SCRIPT_SCHEMA(sequence)
+    script_obj = script.Script(hass, sequence, script_mode=script_mode)
+    wait_started_flag = async_watch_for_action(script_obj, "wait")
 
-    @callback
-    def record_event(event):
-        """Add recorded event to set."""
-        events.append(event)
-
-    hass.bus.async_listen(event, record_event)
-
-    @callback
-    def wait_started_cb():
-        wait_started_flag.set()
-
-    hass.states.async_set("switch.test", "on")
-
-    timeout = timedelta(milliseconds=10)
-    schema = cv.SCRIPT_SCHEMA(
-        [
-            {
-                "wait_template": "{{ states.switch.test.state == 'off' }}",
-                "timeout": timeout,
-            },
-            {"event": event},
-        ]
-    )
-
-    for run_mode in _ALL_RUN_MODES:
-        events = []
-        wait_started_flag.clear()
-
-        if run_mode is None:
-            script_obj = script.Script(hass, schema, change_listener=wait_started_cb)
-        else:
-            script_obj = script.Script(
-                hass, schema, change_listener=wait_started_cb, run_mode=run_mode
-            )
+    try:
+        hass.states.async_set("switch.test", "on")
+        hass.async_create_task(script_obj.async_run())
+        await asyncio.wait_for(wait_started_flag.wait(), 1)
 
-        try:
-            if run_mode == "background":
-                await script_obj.async_run()
-            else:
-                hass.async_create_task(script_obj.async_run())
-            await asyncio.wait_for(wait_started_flag.wait(), 1)
-
-            assert script_obj.is_running
-            assert len(events) == 0
-        except (AssertionError, asyncio.TimeoutError):
-            await script_obj.async_stop()
-            raise
-        else:
-            if run_mode in (None, "legacy"):
-                future = dt_util.utcnow() + timeout
-                async_fire_time_changed(hass, future)
-            await hass.async_block_till_done()
+        assert script_obj.is_running
+        assert len(events) == 0
+    except (AssertionError, asyncio.TimeoutError):
+        await script_obj.async_stop()
+        raise
+    else:
+        async_fire_time_changed(hass, dt_util.utcnow() + timedelta(seconds=5))
+        await hass.async_block_till_done()
 
-            assert not script_obj.is_running
-            assert len(events) == 1
+        assert not script_obj.is_running
+        assert len(events) == n_events
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5164426)